### PR TITLE
Feature: added parquet sampling

### PR DESCRIPTION
**Pull request by menglinw**
parquet sampling function developed in data_utils.py; 
Added sample_nows argument in ParquetData class; 
Added test_len_sampled_data in test_parquet_data.py

### CODE REVIEW

**Comment by taylorfturner** [comment]
@menglinw two things:
- you'll want to open the PR into `dev` 
- also you'll want to rebase onto `dev` to resolve the conflict with `dataprofiler/data_readers/data_utils.py`

File: dataprofiler/data_readers/data_utils.py [hunk 1]
```diff
@@ -439,52 +441,117 @@ def read_csv_df(
...
+    mixed_and_unicode_cols = types[types == "unicode"].index.union(
+        types[types == "mixed"].index
+    )
+
+    for col in mixed_and_unicode_cols:
+        out[col] = out[col].apply(
+            lambda x: x.encode("utf-8").strip() if isinstance(x, str) else x
+        )
+        out[col] = out[col].apply(
+            lambda x: x.decode("utf-8").strip() if isinstance(x, bytes) else x
```
Comment (micdavis) [inline dataprofiler/data_readers/data_utils.py:491]: Do you need the `else x`? It seems like it does nothing.

File: dataprofiler/data_readers/data_utils.py [hunk 2]
```diff
@@ -439,52 +441,117 @@ def read_csv_df(
...
 
-        data_row_df = parquet_file.read_row_group(i).to_pandas()
+    # sample
+    n_rows = table.num_rows
+    if n_rows > sample_nrows:
+        select = np.array([False] * n_rows)
+        select[random.sample(range(n_rows), sample_nrows)] = True
+    else:
+        select = np.array([True] * n_rows)
+    out = table.filter(select).to_pandas()
```
Comment (micdavis) [inline dataprofiler/data_readers/data_utils.py:477]: can `out` be changed to something more descriptive

File: dataprofiler/data_readers/data_utils.py [hunk 3]
```diff
@@ -439,52 +441,117 @@ def read_csv_df(
...
+    if selected_columns:
+        table = pq.read_table(file_path, columns=selected_columns)
+    else:
+        table = pq.read_table(file_path)
 
-        data_row_df = parquet_file.read_row_group(i).to_pandas()
+    # sample
+    n_rows = table.num_rows
+    if n_rows > sample_nrows:
+        select = np.array([False] * n_rows)
```
Comment (micdavis) [inline dataprofiler/data_readers/data_utils.py:473]: can `select` be changed to something more descriptive?

File: dataprofiler/data_readers/data_utils.py [hunk 4]
```diff
@@ -439,52 +441,117 @@ def read_csv_df(
...
 
-        if selected_columns:
-            data_row_df = data_row_df[selected_columns]
+            for col in mixed_and_unicode_cols:
+                data_row_df[col] = data_row_df[col].apply(
+                    lambda x: x.encode("utf-8").strip() if isinstance(x, str) else x
+                )
+                data_row_df[col] = data_row_df[col].apply(
+                    lambda x: x.decode("utf-8").strip() if isinstance(x, bytes) else x
+                )
```
Comment (micdavis) [inline dataprofiler/data_readers/data_utils.py:536]: lots of repeated code from the above function, I wonder if this can be pulled out into it's own function. That way we can just do a function call here and we don't need to repeat code. Just a thought, not a hard ask.

File: dataprofiler/data_readers/data_utils.py [hunk 5]
```diff
@@ -439,52 +441,117 @@ def read_csv_df(
...
+    else:
+        select = np.array([True] * n_rows)
+    out = table.filter(select).to_pandas()
 
-        # Convert all the unicode columns to utf-8
-        types = data_row_df.apply(
-            lambda x: pd.api.types.infer_dtype(x.values, skipna=True)
-        )
+    # Convert all the unicode columns to utf-8
+    types = out.apply(lambda x: pd.api.types.infer_dtype(x.values, skipna=True))
```
Comment (taylorfturner) [inline dataprofiler/data_readers/data_utils.py:480]: I'd vote for improved naming than `types` since that is a fairly native variable to python as a language ... could be a bit more descriptive to what is output from the `.apply()` function

File: dataprofiler/data_readers/data_utils.py (hunk 1 above)
Comment (taylorfturner) [inline dataprofiler/data_readers/data_utils.py:491]:
```suggestion
    for iter_column in mixed_and_unicode_cols:
        out[iter_column] = out[iter_column].apply(
            lambda x: x.encode("utf-8").strip() if isinstance(x, str) else x
        )
        out[iter_column] = out[iter_column].apply(
            lambda x: x.decode("utf-8").strip() if isinstance(x, bytes) else x
```

bit more descriptive

Comment (taylorfturner) [reply dataprofiler/data_readers/data_utils.py:491]: I think this code could be a dry-er too: 
- at its cored these lines of code are 99% the same 
- the only difference is the test of `isinstance(str)` and `isinstance(bytes)`, but at the end of the day they do the same thing even if it is a string of a byte...

I think (you'll want to validate) this could be simplified down to just one line 
```
out[col] = out[col].apply(lambda x: x.encode("utf-8").strip())
```

File: dataprofiler/data_readers/data_utils.py (hunk 4 above)
Comment (taylorfturner) [reply dataprofiler/data_readers/data_utils.py:536]: yeah, I agree @micdavis -- I did just take a look at `main` code in `data_utils.py` and its interesting because that was part of the original release so we do know it works, but also see opportunity to improve. 

Since it was "original release" code, I'm wondering to hold off changing in this PR actually. Thoughts? Follow-up PR @micdavis  (CC @menglinw)

File: dataprofiler/tests/data_readers/test_parquet_data.py [hunk 6]
```diff
@@ -130,6 +150,24 @@ def test_data_formats(self):
...
+        Determine if the parquet file data_formats can be used when sampled
+        """
+        for input_file in self.file_or_buf_list:
+            input_data_obj = Data(input_file["path"], options={"sample_nrows": 100})
+            for data_format in list(input_data_obj._data_formats.keys()):
+                input_data_obj.data_format = data_format
+                self.assertEqual(input_data_obj.data_format, data_format)
+                data = input_data_obj.data
+                if data_format == "dataframe":
+                    import pandas as pd
```
Comment (taylorfturner) [inline dataprofiler/tests/data_readers/test_parquet_data.py:164]: can we not put the `import...` at the top of the file?

**Review by taylorfturner — COMMENTED** [review]
could comments / thoughts

File: dataprofiler/data_readers/data_utils.py [hunk 7]
```diff
@@ -439,52 +441,114 @@ def read_csv_df(
...
     :type file_path: str
+    :param sample_nrows: number of rows being sampled
+    :type sample_nrows: int
+    :param selected_columns: columns need to be read
+    :type selected_columns: list
+    :param read_in_string: return as string type
+    :type read_in_string: bool
     :return:
-    :rtype: Iterator(pd.DataFrame)
+    :rtype:
```
[the remark this answers is not in the document]
Comment (menglinw) [reply dataprofiler/data_readers/data_utils.py]: added return type

File: dataprofiler/data_readers/parquet_data.py [hunk 8]
```diff
@@ -83,6 +84,11 @@ def selected_columns(self) -> List[str]:
         """Return selected columns."""
         return self._selected_columns
 
+    @property
+    def sample_nrows(self) -> Optional[int]:
+        """Return sample_nrows."""
+        return self._sample_nrows
+
```
[the remark this answers is not in the document]
Comment (menglinw) [reply dataprofiler/data_readers/parquet_data.py:91]: thanks!

File: dataprofiler/data_readers/data_utils.py (hunk 2 above)
Comment (menglinw) [reply dataprofiler/data_readers/data_utils.py]: updated  `out` to `sample_df`

File: dataprofiler/data_readers/data_utils.py (hunk 4 above)
Comment (menglinw) [reply dataprofiler/data_readers/data_utils.py]: pulled out as a function `convert_unicode_col_to_utf8` to reduce repeating code,. If we want to holding the change, ready to change it back.

File: dataprofiler/data_readers/data_utils.py [hunk 9]
```diff
@@ -439,52 +441,124 @@ def read_csv_df(
...
+    )
+
+    for col in mixed_and_unicode_cols:
+        input_df[col] = input_df[col].apply(
+            lambda x: x.encode("utf-8").strip() if isinstance(x, str) else x
+        )
+        input_df[col] = input_df[col].apply(
+            lambda x: x.decode("utf-8").strip() if isinstance(x, bytes) else x
+        )
+    return input_df
```
Comment (taylorfturner) [inline dataprofiler/data_readers/data_utils.py:467]: I think this could be a bit more simple code here .... a bit DRY-er

File: dataprofiler/data_readers/data_utils.py (hunk 5 above)
Comment (menglinw) [reply dataprofiler/data_readers/data_utils.py]: changed `types` to `input_column_types` to be more descriptive

File: dataprofiler/data_readers/data_utils.py (hunk 1 above)
Comment (menglinw) [reply dataprofiler/data_readers/data_utils.py]: fixed, but not commit yet

File: dataprofiler/data_readers/data_utils.py (hunk 9 above)
Comment (menglinw) [reply dataprofiler/data_readers/data_utils.py:467]: changed locally, but need to discuss before commit

File: dataprofiler/data_readers/data_utils.py (hunk 1 above)
Comment (menglinw) [reply dataprofiler/data_readers/data_utils.py]: For now, I am leaving this formatting part unchanged, in case causing unknown impact on other parts

File: dataprofiler/data_readers/data_utils.py (hunk 5 above)
Comment (menglinw) [reply dataprofiler/data_readers/data_utils.py]: committed

File: dataprofiler/data_readers/data_utils.py (hunk 1 above)
Comment (menglinw) [reply dataprofiler/data_readers/data_utils.py]: committed

File: dataprofiler/tests/data_readers/test_parquet_data.py (hunk 6 above)
Comment (menglinw) [reply dataprofiler/tests/data_readers/test_parquet_data.py]: fixed

File: dataprofiler/data_readers/data_utils.py (hunk 9 above)
Comment (menglinw) [reply dataprofiler/data_readers/data_utils.py]: leaving this part unchanged after meeting

File: dataprofiler/data_readers/data_utils.py [hunk 10]
```diff
@@ -439,52 +441,130 @@ def read_csv_df(
...
+    if sample_nrows is None:
+        parquet_file = pq.ParquetFile(file_path)
+        data = pd.DataFrame()
+        for i in range(parquet_file.num_row_groups):
+
+            data_row_df = parquet_file.read_row_group(i).to_pandas()
 
-    return data, original_df_dtypes
+            # Convert all the unicode columns to utf-8
+            data_row_df = convert_unicode_col_to_utf8(data_row_df)
```
[the remark this answers is not in the document]
Comment (menglinw) [reply dataprofiler/data_readers/data_utils.py:549]: Yes, this is still needed I think. This `convert_unicode_col_to_utf8()` here is used to format dataframe when we need to read all data, while the `convert_unicode_col_to_utf8()`  in `sample_parquet()` is used to format sampled dataframe, so the output from both (sampled or all data) can in the same format.

File: dataprofiler/tests/data_readers/test_parquet_data.py [hunk 11]
```diff
@@ -112,6 +119,15 @@ def test_reload_data(self):
...
             self.assertEqual(input_file["path"], input_data_obj.input_file_path)
 
+            input_data_obj_sampled = Data(
+                input_file["path"], options={"sample_nrows": 100}
+            )
+            input_data_obj_sampled.reload(
+                input_file["path"], options={"sample_nrows": 100}
+            )
+            self.assertEqual(input_data_obj_sampled.data_type, "parquet")
+            self.assertEqual(input_file["path"], input_data_obj_sampled.input_file_path)
```
[the remark this answers is not in the document]
Comment (menglinw) [reply dataprofiler/tests/data_readers/test_parquet_data.py]: I agree. Checking before and after reload with sampling enabled is more comprehensive . I have edited based on your suggested code and committed the change. Thanks for your feedbacks!

**Review by micdavis — APPROVED** [review]

File: dataprofiler/data_readers/avro_data.py [hunk 12]
```diff
@@ -67,7 +67,7 @@ def _load_data_from_file(self, input_file_path: str) -> List:
             # even when the option encoding='utf-8' is added. It may come from
             # some special compression codec, e.g., snappy. Then, binary mode
             # reading is currently used to get the dict-formatted lines.
-            df_reader: fastavro.reader = fastavro.reader(input_file)
+            df_reader = fastavro.reader(input_file)
```
Comment (taylorfturner) [inline dataprofiler/data_readers/avro_data.py:70]: what I still don't understand, yet, is why this failed all of the sudden.... lol

Comment (menglinw) [reply dataprofiler/data_readers/avro_data.py:70]: Lol, I am having the same question... It was working perfectly at my last commit, and suddenly, not anymore..

File: requirements.txt [hunk 13]
```diff
@@ -6,7 +6,7 @@ python-dateutil>=2.7.5
 pytz>=2020.1
 pyarrow>=1.0.1
 chardet>=3.0.4
-fastavro>=1.0.0.post1
+fastavro>=1.0.0.post1,<1.9.0
```
Comment (taylorfturner) [inline requirements.txt:9]: so probably don't need this anymore -- let's remove this

**Review by taylorfturner — CHANGES_REQUESTED** [review]
let's update the requirements.tct

File: requirements.txt (hunk 13 above)
Comment (menglinw) [reply requirements.txt]: I agree. Changed back

File: dataprofiler/data_readers/avro_data.py (hunk 12 above)
Comment (menglinw) [reply dataprofiler/data_readers/avro_data.py:70]: Sure, happy to do that! It may take us some time to find where the error comes from, but fixing it is exciting

File: dataprofiler/tests/data_readers/test_parquet_data.py [hunk 14]
```diff
@@ -123,13 +154,24 @@ def test_data_formats(self):
...
+            for data_format in list(input_data_obj_sampled._data_formats.keys()):
+                input_data_obj_sampled.data_format = data_format
+                self.assertEqual(input_data_obj_sampled.data_format, data_format)
+                data_sampled = input_data_obj_sampled.data
+                if data_format == "dataframe":
+                    self.assertIsInstance(data_sampled, pd.DataFrame)
+                elif data_format in ["records", "json"]:
+                    self.assertIsInstance(data_sampled, list)
+                    self.assertIsInstance(data_sampled[0], str)
+
```
Comment (taylorfturner) [inline dataprofiler/tests/data_readers/test_parquet_data.py:174]: this reads a little verbose.... wondering if needed since its adding an option to the test...

File: dataprofiler/tests/data_readers/test_parquet_data.py [hunk 15]
```diff
@@ -102,6 +104,11 @@ def test_specifying_data_type(self):
             input_data_obj = Data(input_file["path"], data_type="parquet")
             self.assertEqual(input_data_obj.data_type, "parquet")
 
+            input_data_obj_sampled = Data(
+                input_file["path"], data_type="parquet", options={"sample_nrows": 100}
+            )
+            self.assertEqual(input_data_obj_sampled.data_type, "parquet")
+
```
[the remark this answers is not in the document]
Comment (menglinw) [reply dataprofiler/tests/data_readers/test_parquet_data.py:111]: agree, seems adding sampling option wouldn't change the type. Deleted.

File: dataprofiler/tests/data_readers/test_parquet_data.py (hunk 14 above)
Comment (menglinw) [reply dataprofiler/tests/data_readers/test_parquet_data.py:174]: I think this is necessary. The new parquet sampling feature is using a different parquet reading method and then format, so I think it might be safer to check if the sampled data still meet our format requirements.

File: dataprofiler/data_readers/data_utils.py [hunk 16]
```diff
@@ -439,52 +441,130 @@ def read_csv_df(
...
+
+def sample_parquet(
     file_path: str,
+    sample_nrows: int,
     selected_columns: Optional[List[str]] = None,
     read_in_string: bool = False,
 ) -> Tuple[pd.DataFrame, pd.Series]:
     """
-    Return an iterator that returns one row group each time.
+    Read parquet file, sample n row from it and return a dataframe.
```
[the remark this answers is not in the document]
Comment (menglinw) [reply dataprofiler/data_readers/data_utils.py]: fixed

**Review by taylorfturner — APPROVED** [review]

File: dataprofiler/tests/data_readers/test_parquet_data.py (hunk 14 above)
Comment (micdavis) [reply dataprofiler/tests/data_readers/test_parquet_data.py:145]: I don't mind too much if tests are verbose, as long as their readable. Better to be verbose than miss edge cases imo.

**Review by micdavis — APPROVED** [review]
